### PR TITLE
Basic CCheckQueue Benchmarks

### PR DESCRIPTION
**Pull request by JeremyRubin**
Some basic benchmarks for the CCheckQueue, for future work to compare to.

### CODE REVIEW

File: src/bench/checkqueue.cpp [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+#include <vector>
+#include <boost/thread/thread.hpp>
+#include "random.h"
+
+
+// This Benchmark tests the CheckQueue with the lightest
+// weight Checks, so it should make any lock contention
+// particularly visible
+static void CCheckQueueSpeed(benchmark::State& state)
+{
```
Comment (ryanofsky) [inline src/bench/checkqueue.cpp]: Maybe declare some constants here with the various magic numbers & things that could be tuned in the benchmark, like size of the queue, number of threads, number of jobs to add, number of times to call add, etc.

File: src/bench/checkqueue.cpp [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    };
+    CCheckQueue<FakeJobNoWork> queue {128};
+    boost::thread_group tg;
+    for (auto x = 0; x < std::max(2, GetNumCores()); ++x) {
+       tg.create_thread([&]{queue.Thread();});
+    }
+    while (state.KeepRunning()) {
+        CCheckQueueControl<FakeJobNoWork> control(&queue);
+        // We can make vChecks out of the loop because calling Add doesn't
+        // change the size of the vector.
```
Comment (ryanofsky) [inline src/bench/checkqueue.cpp]: Maybe add an assert to the end of the benchmark to check this? It's be easy to imagine the vector being moved from in the future, which would invalidate the benchmark.

Comment (ryanofsky) [inline src/bench/checkqueue.cpp]: Also, if this is true, any reason not to move vChecks declaration out of the while loop too?

File: src/bench/checkqueue.cpp [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+        std::vector<FakeJobNoWork> vChecks;
+        vChecks.resize(30);
```
Comment (ryanofsky) [inline src/bench/checkqueue.cpp]: Could call the vector constructor taking a size to eliminate the resize line.

File: src/bench/checkqueue.cpp [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+        void swap(PrevectorJob& x){p.swap(x.p);};
+    };
+    CCheckQueue<PrevectorJob> queue {128};
+    boost::thread_group tg;
+    for (auto x = 0; x < std::max(2, GetNumCores()); ++x) {
+       tg.create_thread([&]{queue.Thread();});
+    }
+    while (state.KeepRunning()) {
+        // Make insecure_rand here so that each iteration is identical.
+        FastRandomContext insecure_rand(true);
```
Comment (ryanofsky) [inline src/bench/checkqueue.cpp]: Maybe move this outside while loop, if whatever time this takes isn't meaningful to the benchmark.

File: src/bench/checkqueue.cpp [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+// where checks all contain a prevector that is indirect 50% of the time
+// and there is a little bit of work done between calls to Add.
+static void CCheckQueueSpeedPrevectorJob(benchmark::State& state)
+{
+    struct PrevectorJob {
+        prevector<28, uint8_t> p;
+        PrevectorJob(){
+        }
+        PrevectorJob(FastRandomContext& insecure_rand){
+            p.resize(insecure_rand.rand32() % 56);
```
Comment (ryanofsky) [inline src/bench/checkqueue.cpp]: Why do this in the constructor instead of the call operator? Wouldn't it be more realistic to have the work done in the worker threads instead of the control thread? Maybe add a comment if this is the intended behavior.

**Review by ryanofsky — APPROVED** [review]
utACK 9f03110f3256e81173c65c2ed1810efaa74fbc5b. Looks fine, just minor comments.

File: src/bench/checkqueue.cpp (hunk 4 above)
Comment (JeremyRubin) [inline src/bench/checkqueue.cpp]: It is meaningful for each iteration to be identical.

File: src/bench/checkqueue.cpp (hunk 5 above)
Comment (JeremyRubin) [inline src/bench/checkqueue.cpp]: FastRandomContext is not thread safe.

Also we do want the main thread to be doing some work while the worker threads process.

File: src/bench/checkqueue.cpp (hunk 4 above)
Comment (ryanofsky) [inline src/bench/checkqueue.cpp]: Oops, should have read the comment above. Though maybe I would reword it to something like "Start with a fresh insecure_rand here" rather than "Make insecure_rand here."

**Review by ryanofsky — APPROVED** [review]
utACK aad4cb50599b15d6fc439e465a616635e96466d1

**Comment by fanquake** [comment]
```
CCheckQueueSpeed,576,0.001812756061554,0.001984249800444,0.001944298545520,6149498,6731089,6595898
CCheckQueueSpeed,576,0.001704573631287,0.001963026821613,0.001897984494766,5782270,6659114,6438475
CCheckQueueSpeed,576,0.001838877797127,0.001967173069715,0.001926153070397,6237919,6675994,6534347
CCheckQueueSpeed,576,0.001697123050690,0.001930065453053,0.001841966890626,5757193,6547343,6248762
CCheckQueueSpeed,576,0.001837179064751,0.001974612474442,0.001906548523241,6232267,6698444,6467841
```
```
CCheckQueueSpeedPrevectorJob,448,0.002283558249474,0.002333313226700,0.002298991062811,7746465,7915155,7799211
CCheckQueueSpeedPrevectorJob,448,0.002310715615749,0.002362936735153,0.002330580460174,7838558,8015741,7906361
CCheckQueueSpeedPrevectorJob,448,0.002285875380039,0.002489343285561,0.002338334918022,7754281,8444561,7932261
CCheckQueueSpeedPrevectorJob,448,0.002281628549099,0.002457186579704,0.002311725169420,7739965,8335454,7842407
CCheckQueueSpeedPrevectorJob,448,0.002286218106747,0.002332687377930,0.002305187284946,7755492,7913158,7820222
```

**Comment by laanwj** [comment]
utack aad4cb5
